Add debugging steps for DiscoverEC2 User Task issues

marcoandredinis opened the pull request:
This PR adds a new field when getting an User Task. Description contains a markdown document that should be displayed to the user which helps them fix the issue with the task.

hugoShaka reviewed (commented):
The code LGTM, my only question is regarding the description location. Do we want to hardcode those in `api/` or should the description be set server-side and live in `lib/`?

marcoandredinis commented:
> The code LGTM, my only question is regarding the description location. Do we want to hardcode those in `api/` or should the description be set server-side and live in `lib/`?

That's a good point. I'm not sure which one is a better fit.
IMO, it's fine in both places.
If you or someone thinks this should be placed under `lib/` I'm fine changing it.

r0mant commented on api/types/usertasks/descriptions/ec2-ssm-agent-connection-lost.md:
```diff
@@ -0,0 +1,27 @@
+Auto enrolling EC2 instances requires the SSM Agent to be installed and running on them.
+Unfortunately, some instances lost connection to Amazon Systems Manager.
```
```suggestion
Some instances appear to have lost connection to Amazon Systems Manager.
```

r0mant commented on api/types/usertasks/descriptions/ec2-ssm-agent-connection-lost.md:
```diff
@@ -0,0 +1,27 @@
+
+Ensure the instance's security groups allows outbound connections to Amazon Systems Manager endpoints.
+Allowing outbound on port 443 is enough for the agent to connect to AWS.
+This might have changed recently.
```
```suggestion
```

r0mant commented on api/types/usertasks/descriptions/ec2-ssm-agent-connection-lost.md:
```diff
@@ -0,0 +1,27 @@
+Allowing outbound on port 443 is enough for the agent to connect to AWS.
+This might have changed recently.
+
+**Missing policies in the instance**
```
```suggestion
**Instance is missing IAM policy**
```

r0mant commented on api/types/usertasks/descriptions/ec2-ssm-agent-connection-lost.md:
```diff
@@ -0,0 +1,27 @@
+The SSM Agent requires the `AmazonSSMManagedInstanceCore` managed policy.
+Ensure the instance has an IAM Profile and that it includes the above policy.
+For more information please refer to [this page](https://docs.aws.amazon.com/systems-manager/latest/userguide/session-manager-getting-started-instance-profile.html).
+This might have changed recently.
```
```suggestion
```

r0mant commented on api/types/usertasks/descriptions/ec2-ssm-agent-not-registered.md:
```diff
@@ -0,0 +1,25 @@
+Auto enrolling EC2 instances requires the SSM Agent to be installed and running on them.
+Unfortunately, some instances are not connecting to Amazon Systems Manager.
```
```suggestion
Some instances failed to connect to Amazon Systems Manager.
```

r0mant commented on api/types/usertasks/descriptions/ec2-ssm-invocation-failure.md:
```diff
@@ -0,0 +1,19 @@
+Teleport failed to access the SSM Agent to auto enroll the instance.
+This is not a failure of the script that ran in the instance, but rather a failure when trying to communicate with the AWS Systems Manager service.
```
```suggestion
Some instances failed to communicate with the AWS Systems Manager service to execute the install script.
```

r0mant commented on api/types/usertasks/descriptions/ec2-ssm-invocation-failure.md:
```diff
@@ -0,0 +1,19 @@
+
+**SSM Document is invalid**
+
+Teleport uses an SSM Document to run an installation script.
```
How/where can a user find this document?

marcoandredinis replied:
I'll the SSM Document name to the list of fields.
UI should then build the SSM Document URL (we have the region and the SSM Document name, so that should be simple).

r0mant replied:
@marcoandredinis I think these messages should be templated using text/template. Then whatever returns these messages to the web UI can parameterize them with appropriate data without having to rely on the web UI to assemble a bunch of stuff.

marcoandredinis replied:
We can template everything or we show have specific fields for metadata and then only the description is templated.

Taking the following screen as example
![image](https://github.com/user-attachments/assets/79bebbdd-1251-49c4-9472-44a000f82a31)

I was thinking that only the `Details` section was to be converted from Markdown into HTML (aka sent over the API).

After reading your comment I wonder if you want it all to be generated in the backend.
Which parts should be rendered in the backend?
Title
Metadata (region, account id)
Details
Impacted instances

We can do it all, if that's what we want.
However, we have to be careful because there are some elements that might look out of place (using a different CSS style) because the library we'll be using will not use the same set of styles that we have.
While this might be easy to fix for things that are mostly text (bolds, lists and such), it might not be easy for more complex elements like tables.

(@kimlisa please correct me if I'm wrong because I don't know the library we'll be using)

michellescripts replied:
React-markdown sounds OK, but we don't actually use it right now since Assist is deprecated. 

Is there a reason to send markdown to the front end, vs sending an error code indicating which display to show, and the UI has a collection of views?

r0mant replied:
@michellescripts The reason I'd prefer for the frontend to just be a simpler renderer in this case is so we can easily expand our "library" of errors (with descriptions and other metadata) without having to update UI every time we want to add a new type of user task.

michellescripts replied:
I'm fine with that! I do agree that templating in links for the user will be a better UX but we can tackle that whenever.

marcoandredinis replied:
We have 4 sections: title, metadata, description and table with affected resources.

My suggestion:
- Title is just a string.
- Metadata should come from the backend and separate fields and rendered in the frontend.
- Description comes from the backend as markdown and is rendered in the frontend using react-markdown.
- Table comes as a list from the backend, and is rendered in the frontend.

r0mant commented on api/types/usertasks/descriptions/ec2-ssm-script-failure.md:
```diff
@@ -0,0 +1,3 @@
+Teleport was able to reach the SSM Agent inside the EC2 instance, however the script returned an error.
```
Also, how can a user see and update the script?
```suggestion
Teleport was able to reach the SSM Agent inside the EC2 instance, however the install script returned an error.
```

marcoandredinis replied:
I'll the InstallerScript used to the list of fields.
UI can suggest the user to do a `tctl edit installer/<script-name>` so that the user can edit the script.
What do you think?

I'm not so sure about this one as I am for the SSM Document, because for this one, the user needs to change tools (we don't expose CRUD for installer scripts in the UI).

I'll add them anyway.

r0mant commented on api/types/usertasks/descriptions/ec2-ssm-script-failure.md:
```diff
@@ -0,0 +1,3 @@
+Teleport was able to reach the SSM Agent inside the EC2 instance, however the script returned an error.
+
+You can click below in the Invocation URL and get further details on why the script failed.
```
The URL seems to be missing?

marcoandredinis replied:
The URL will appear below.
We have the the description and then a table with affected instances
That table will contain, for each instance, its invocation URL.

r0mant commented on api/types/usertasks/descriptions/ec2-ssm-unsupported-os.md:
```diff
@@ -0,0 +1,9 @@
+Auto enrolling EC2 instances requires a compatible Operating System.
+
+Teleport only supports Linux instances when auto-enrolling them into the cluster.
+
+<!---
+We might suggest users to use specific tags on the instances they want to enroll and then change the DiscoveryConfig's matcher to only include those tags.
+This is only possible using `tctl edit discovery_config/....` which requires a tool/context switch.
+When we include DiscoveryConfigs in the UI, we can teach the user how to do it.
+-->
```
```suggestion
```

r0mant reviewed (commented):
With some comments.

hugoShaka reviewed (approved):
LGTM

I'd prefer the description to be set server-side and live `lib/` but that's not blocking.

marcoandredinis committed:
Add debugging steps for DiscoverEC2 User Task issues

This PR adds a new field when getting an User Task.
Description contains a markdown document that should be displayed to the
user which helps them fix the issue with the task.

marcoandredinis committed:
move descriptions from api to lib

marcoandredinis commented:
> LGTM
> 
> I'd prefer the description to be set server-side and live `lib/` but that's not blocking.

Done 👍

marcoandredinis commented:
@r0mant I've added two new fields: SSM Document and Installer Script

We can show them in the UI.
For the SSM Document we can link out to Amazon Console (we have region and the document name, and that's all it's needed to build the URL)
For the Installer Script we can teach the user how to do it with `tctl` for now (while we don't expose that resource in the UI).

Can you please take another look?

kimlisa commented:
> @kimlisa I know you were looking for a way to render this markdown in the UI. Did you reach any conclusion? Can we pursue this method or should we think of some other way to show those instructions?

yeah, assist uses `react-markdown` and we'll do the same

r0mant commented on lib/usertasks/descriptions/ec2-ssm-invocation-failure.md:
```diff
@@ -0,0 +1,19 @@
+
+**SSM Document is invalid**
+
+Teleport uses an SSM Document to run an installation script.
+If the document is changed or removed, it might no longer work.
```
It is still not clear where this SSM document or install script is. I think we need to template these messages like in my other comment to make them more useful to the user.

r0mant reviewed (commented):
@marcoandredinis @kimlisa Before merging this, I'd like to see how these are presented in the web UI so we can see if things make sense or need more tweaks to be more clear and actionable. I understand this depends on having this view implemented in the web UI so I think it's ok to hold this until we have it.

marcoandredinis commented:
@r0mant Can we merge this one?

I think we came to a conclusion in this discussion thread
https://github.com/gravitational/teleport/pull/47513#discussion_r1875834404

r0mant reviewed (approved):
@marcoandredinis Yeah let's get this in.
